File: backend/app/schemas/agent_schemas.py
"""
Agent Schemas
=============
Request/response contracts for the autonomous agent planner.

These are deliberately separate from the security and llm schemas: the agent
runtime is an isolated subsystem and changes here must never ripple into the
manual "Scan Page" or chat flows.
"""

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


class AgentSemanticElement(BaseModel):
    """One actionable element, as produced by the frontend State Builder."""

    id: str
    role: str
    name: str = ""
    value: Optional[str] = None
    description: Optional[str] = None
    url: Optional[str] = None
    disabled: Optional[bool] = None
    required: Optional[bool] = None
    focused: Optional[bool] = None
    expanded: Optional[bool] = None
    selected: Optional[bool] = None
    checked: Optional[str] = None
    invalid: Optional[str] = None
    nearbyText: Optional[str] = None
    placeholder: Optional[str] = None
    container: Optional[str] = None
    inputType: Optional[str] = None
    nameAttr: Optional[str] = None


class AgentSemanticDialog(BaseModel):
    id: str
    role: str
    name: str = ""
    modal: bool = False


class AgentValidationIssue(BaseModel):
    elementId: str
    role: str
    message: str = ""


class AgentPageState(BaseModel):
    """The compact semantic state. The raw accessibility tree never arrives here."""

    url: str = ""
    title: str = ""
    elements: List[AgentSemanticElement] = Field(default_factory=list)
    focusedElementId: Optional[str] = None
    dialogs: List[AgentSemanticDialog] = Field(default_factory=list)
    validationErrors: List[AgentValidationIssue] = Field(default_factory=list)
    selectedElementIds: List[str] = Field(default_factory=list)
    truncated: bool = False


class AgentStepRecord(BaseModel):
    tool: str
    summary: str = ""
    succeeded: bool = True


class AgentFailureRecord(BaseModel):
    tool: str
    reason: str = ""
    code: Optional[str] = None


class AgentWorkingMemory(BaseModel):
    """
    The planner's entire recollection. Deliberately not a conversation
    transcript: a bounded, structured summary keeps the prompt small and stops
    older page content from accumulating in context.
    """

    goal: str = ""
    completed_steps: List[AgentStepRecord] = Field(default_factory=list)
    pending_steps: List[str] = Field(default_factory=list)
    failures: List[AgentFailureRecord] = Field(default_factory=list)
    retries: int = 0
    current_page: str = ""


class AgentPlanRequest(BaseModel):
    goal: str
    working_memory: AgentWorkingMemory = Field(default_factory=AgentWorkingMemory)
    page_state: AgentPageState = Field(default_factory=AgentPageState)


class AgentToolCall(BaseModel):
    """A single validated action. `arguments` is already schema-checked."""

    tool: str
    arguments: Dict[str, Any] = Field(default_factory=dict)
    """True when the runtime must obtain explicit human consent first."""
    requires_approval: bool = False


class AgentPageSnapshot(BaseModel):
    """
    Deep CDP capture of the page the agent is about to act on.

    Intentionally a separate model from `WebpageCheckRequest`, even though the
    field names overlap. The manual scan contract and the agent contract must
    be able to change independently — that isolation is the whole point.
    """

    visible_text: str = ""
    hidden_text: str = ""
    html_comments: str = ""
    meta_tags: str = ""
    input_values: str = ""
    aria_text: str = ""
    iframe_content: str = ""
    shadow_dom_content: str = ""
    inline_javascript: str = ""
    external_javascript: str = ""
    css_content: str = ""
    css_generated_content: str = ""
    network_responses: str = ""
    websocket_messages: str = ""
    service_worker_activity: str = ""
    dom_snapshot_content: str = ""
    page_title: str = ""
    url: str = ""


class AgentScanRequest(BaseModel):
    """A scan request for one agent iteration."""

    task_id: str
    url: str = ""
    page_hash: str = ""
    snapshot: AgentPageSnapshot


class AgentThreatFindingModel(BaseModel):
    source: str
    confidence: float
    matched_patterns: List[str] = Field(default_factory=list)
    matched_evidence: List[str] = Field(default_factory=list)
    excerpt: str = ""


class AgentScanResponse(BaseModel):
    """
    The machine decision that gates execution.

    `allowed == False` trips the runtime circuit breaker: the pending planner
    action is discarded and the task ends.
    """

    allowed: bool
    task_id: str
    url: str = ""
    page_hash: str = ""
    risk_level: Literal["low", "medium", "high"]
    confidence: float
    summary_reason: str
    matched_patterns: List[str] = Field(default_factory=list)
    blocked_sources: List[str] = Field(default_factory=list)
    findings: List[AgentThreatFindingModel] = Field(default_factory=list)
    scanned_chunks: int = 0
    classifier_mode: str = ""
    scanned_at: str = ""


class AgentPlanResponse(BaseModel):
    """
    The next action, or a short queue of them.

    `tool_call` is the first queued action, kept so a caller that only ever
    executes one action at a time needs no change.
    """

    tool_calls: List[AgentToolCall] = Field(min_length=1)
    tool_call: AgentToolCall
    confidence: float = Field(ge=0.0, le=1.0)
    """True when confidence fell below the configured threshold and the runtime
    should pause for the user instead of executing."""
    needs_user_confirmation: bool = False
    reason: str = ""
    model: str = ""
    planner_mode: Literal["llm"] = "llm"

File: backend-node/src/schemas/agent.schemas.ts
// Port of backend/app/schemas/agent_schemas.py.
// Deliberately separate from security.schemas.ts, even though field names
// overlap: the manual-scan contract and the agent contract must be able to
// change independently — that isolation is the whole point.
//
// Phase 3 ported the /agent/scan-active-page + /agent/security/events +
// /agent/tools slice below. Phase 4 adds AgentPlanRequest/AgentPlanResponse
// and the working-memory/page-state schemas the LLM-backed planner consumes.
import { Type, type Static } from '@sinclair/typebox'

// Every field defaults to "" in the Python model — all optional here too.
export const AgentPageSnapshotSchema = Type.Object({
  visible_text: Type.Optional(Type.String()),
  hidden_text: Type.Optional(Type.String()),
  html_comments: Type.Optional(Type.String()),
  meta_tags: Type.Optional(Type.String()),
  input_values: Type.Optional(Type.String()),
  aria_text: Type.Optional(Type.String()),
  iframe_content: Type.Optional(Type.String()),
  shadow_dom_content: Type.Optional(Type.String()),
  inline_javascript: Type.Optional(Type.String()),
  external_javascript: Type.Optional(Type.String()),
  css_content: Type.Optional(Type.String()),
  css_generated_content: Type.Optional(Type.String()),
  network_responses: Type.Optional(Type.String()),
  websocket_messages: Type.Optional(Type.String()),
  service_worker_activity: Type.Optional(Type.String()),
  dom_snapshot_content: Type.Optional(Type.String()),
  page_title: Type.Optional(Type.String()),
  url: Type.Optional(Type.String()),
  // Node-only extended channels (Phase 2) — see security.routes.ts. Accepted
  // here too since the same CDP capture is reused for both scan paths.
  source_maps: Type.Optional(Type.String()),
  redirects: Type.Optional(Type.String()),
  third_party_resources: Type.Optional(Type.String()),
  suspicious_domains: Type.Optional(Type.String()),
  frame_navigation: Type.Optional(Type.String()),
  runtime_script_activity: Type.Optional(Type.String()),
  loaded_resources: Type.Optional(Type.String()),
})

export const AgentScanRequestSchema = Type.Object({
  task_id: Type.String(),
  url: Type.Optional(Type.String()),
  page_hash: Type.Optional(Type.String()),
  snapshot: AgentPageSnapshotSchema,
})

export const AgentThreatFindingSchema = Type.Object({
  source: Type.String(),
  confidence: Type.Number(),
  matched_patterns: Type.Array(Type.String()),
  matched_evidence: Type.Array(Type.String()),
  excerpt: Type.String(),
})

export const AgentScanResponseSchema = Type.Object({
  allowed: Type.Boolean(),
  task_id: Type.String(),
  url: Type.String(),
  page_hash: Type.String(),
  risk_level: Type.Union([Type.Literal('low'), Type.Literal('medium'), Type.Literal('high')]),
  confidence: Type.Number(),
  summary_reason: Type.String(),
  matched_patterns: Type.Array(Type.String()),
  blocked_sources: Type.Array(Type.String()),
  findings: Type.Array(AgentThreatFindingSchema),
  scanned_chunks: Type.Integer(),
  classifier_mode: Type.String(),
  scanned_at: Type.String(),
})

export type AgentPageSnapshot = Static<typeof AgentPageSnapshotSchema>
export type AgentScanRequest = Static<typeof AgentScanRequestSchema>
export type AgentScanResponse = Static<typeof AgentScanResponseSchema>

// ─────────────────────────── Phase 4: planner contract ───────────────────────────
// One actionable element, as produced by the frontend State Builder.
export const AgentSemanticElementSchema = Type.Object({
  id: Type.String(),
  role: Type.String(),
  name: Type.Optional(Type.String()),
  value: Type.Optional(Type.Union([Type.String(), Type.Null()])),
  description: Type.Optional(Type.Union([Type.String(), Type.Null()])),
  url: Type.Optional(Type.Union([Type.String(), Type.Null()])),
  disabled: Type.Optional(Type.Union([Type.Boolean(), Type.Null()])),
  required: Type.Optional(Type.Union([Type.Boolean(), Type.Null()])),
  focused: Type.Optional(Type.Union([Type.Boolean(), Type.Null()])),
  expanded: Type.Optional(Type.Union([Type.Boolean(), Type.Null()])),
  selected: Type.Optional(Type.Union([Type.Boolean(), Type.Null()])),
  checked: Type.Optional(Type.Union([Type.String(), Type.Null()])),
  invalid: Type.Optional(Type.Union([Type.String(), Type.Null()])),
  nearbyText: Type.Optional(Type.Union([Type.String(), Type.Null()])),
  placeholder: Type.Optional(Type.Union([Type.String(), Type.Null()])),
  container: Type.Optional(Type.Union([Type.String(), Type.Null()])),
  inputType: Type.Optional(Type.Union([Type.String(), Type.Null()])),
  nameAttr: Type.Optional(Type.Union([Type.String(), Type.Null()])),
})

export const AgentSemanticDialogSchema = Type.Object({
  id: Type.String(),
  role: Type.String(),
  name: Type.Optional(Type.String()),
  modal: Type.Optional(Type.Boolean()),
})

export const AgentValidationIssueSchema = Type.Object({
  elementId: Type.String(),
  role: Type.String(),
  message: Type.Optional(Type.String()),
})

// The compact semantic state. The raw accessibility tree never arrives here.
export const AgentPageStateSchema = Type.Object({
  url: Type.Optional(Type.String()),
  title: Type.Optional(Type.String()),
  elements: Type.Optional(Type.Array(AgentSemanticElementSchema)),
  focusedElementId: Type.Optional(Type.Union([Type.String(), Type.Null()])),
  dialogs: Type.Optional(Type.Array(AgentSemanticDialogSchema)),
  validationErrors: Type.Optional(Type.Array(AgentValidationIssueSchema)),
  selectedElementIds: Type.Optional(Type.Array(Type.String())),
  truncated: Type.Optional(Type.Boolean()),
})

export const AgentStepRecordSchema = Type.Object({
  tool: Type.String(),
  summary: Type.Optional(Type.String()),
  succeeded: Type.Optional(Type.Boolean()),
})

export const AgentFailureRecordSchema = Type.Object({
  tool: Type.String(),
  reason: Type.Optional(Type.String()),
  code: Type.Optional(Type.Union([Type.String(), Type.Null()])),
})

// The planner's entire recollection. Deliberately not a conversation
// transcript: a bounded, structured summary keeps the prompt small and stops
// older page content from accumulating in context.
export const AgentWorkingMemorySchema = Type.Object({
  goal: Type.Optional(Type.String()),
  completed_steps: Type.Optional(Type.Array(AgentStepRecordSchema)),
  pending_steps: Type.Optional(Type.Array(Type.String())),
  failures: Type.Optional(Type.Array(AgentFailureRecordSchema)),
  retries: Type.Optional(Type.Integer()),
  current_page: Type.Optional(Type.String()),
})

export const AgentPlanRequestSchema = Type.Object({
  goal: Type.String(),
  working_memory: Type.Optional(AgentWorkingMemorySchema),
  page_state: Type.Optional(AgentPageStateSchema),
})

// A single validated action. `arguments` is already schema-checked.
export const AgentToolCallSchema = Type.Object({
  tool: Type.String(),
  arguments: Type.Record(Type.String(), Type.Unknown()),
  requires_approval: Type.Boolean(),
})

export const AgentPlanResponseSchema = Type.Object({
  tool_calls: Type.Array(AgentToolCallSchema, { minItems: 1 }),
  tool_call: AgentToolCallSchema,
  confidence: Type.Number({ minimum: 0, maximum: 1 }),
  needs_user_confirmation: Type.Boolean(),
  reason: Type.String(),
  model: Type.String(),
  planner_mode: Type.Literal('llm'),
})

export type AgentSemanticElement = Static<typeof AgentSemanticElementSchema>
export type AgentSemanticDialog = Static<typeof AgentSemanticDialogSchema>
export type AgentValidationIssue = Static<typeof AgentValidationIssueSchema>
export type AgentPageState = Static<typeof AgentPageStateSchema>
export type AgentStepRecord = Static<typeof AgentStepRecordSchema>
export type AgentFailureRecord = Static<typeof AgentFailureRecordSchema>
export type AgentWorkingMemory = Static<typeof AgentWorkingMemorySchema>
export type AgentPlanRequest = Static<typeof AgentPlanRequestSchema>
export type AgentToolCall = Static<typeof AgentToolCallSchema>
export type AgentPlanResponse = Static<typeof AgentPlanResponseSchema>
